File: problems/Lowest Common Ancestor of Deepest Leaves/lcaDeepestLeaves.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode LcaDeepestLeaves(TreeNode root) {
        var (node, _)  = lca(root, 0);
        return node;
    }

    public (TreeNode, int) lca(TreeNode root, int depth) {
        if (null == root || (null == root.left && null == root.right)) {
            return (root, depth);
        }

        (TreeNode left, int leftDepth) = lca(root.left, 1 + depth);
        (TreeNode right, int rightDepth) = lca(root.right, 1 + depth);

        if (null == left && null == right) {
            return (root, depth);
        }

        if (null == right || (leftDepth > rightDepth)) {
            return (left, leftDepth);
        }

        if (null == left || (rightDepth > leftDepth)) {
            return (right, rightDepth);
        }

        return (root, leftDepth);
    }
}

File: problems/Lowest Common Ancestor of Deepest Leaves/lcaDeepestLeaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        Queue<TreeNode> bfs = new LinkedList<>();
        List<TreeNode> leaves = new ArrayList<>();

        if (null != root) {
            bfs.add(root);
            leaves.add(root);
        }

        while (!bfs.isEmpty()) {
            int count = bfs.size();
            leaves.clear();

            for (int idx = 0; count > idx; ++idx) {
                TreeNode node = bfs.poll();
                leaves.add(node);

                if (null != node.left) {
                    bfs.add(node.left);
                }

                if (null != node.right) {
                    bfs.add(node.right);
                }
            }
        }

        if (1 == leaves.size()) {
            return leaves.get(0);
        }

        return lowestCommonAncestor(root, leaves.get(0), leaves.get(leaves.size() - 1));
    }

    private TreeNode lca = null;

    private TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        dfs(root, p, q);

        return this.lca;
    }

    private boolean dfs(TreeNode node, TreeNode p, TreeNode q) {
        if (null == node) {
            return false;
        }

        int left = dfs(node.left, p, q) ? 1 : 0;
        int mid = node.val == p.val || node.val == q.val ? 1 : 0;
        int right = dfs(node.right, p, q) ? 1 : 0;

        if (2 == left + mid + right) {
            this.lca = node;
        }

        return 1 == left + mid + right;
    }
}
